### DISCUSSION TITLE
fStorage.GetSynchronous Thread Save

### DISCUSSION THREAD

**Discussion by itsfintom** [SchneiderInfosystems/FB4D#120] [Q&A] [answered]
Is the fStorage.GetSynchronous implemented thread safe ?
I want to run the GetSynchronous and DownloadToStream in a Helper thread, because it would fit better to the overall design of the application.
I tried it and it runs fine..But yeah that does not mean it is designed for that.
It could be that we are talking about 5-10 concurrent running operations.
If this is not designed for it, I would need to implement a scheduler thread to only run one sync operation at a time.

Just a short quick hint on that would be nice.

Thank you
Thomas

**Comment by SchneiderInfosystems** [comment, marked as answer]
Hi Thomas

No, in the current solution this is not thread-safe implemented. The access to fStorageObjs in line 268 would need to be protected by a critical section.

The current library expects that all synchronous calls to be called from the same background thread while all async calls are called from the GUI thread.

If you attend EKON 25 in Düsseldorf (Germany) next week, we could discuss such ideas for future changes.

Br, 
Christoph

**Reply by itsfintom** [reply]
Yes,..
What I have done is the following.
For each worker task I create a new IFirebaseStorage for the bucket and fAuth.
This is private to the task.
To limit the tasks, I have introduced a Semaphore, to let only run a specific number of this worker tasks in parallel.
The Semaphore was needed because on iOS, the OS limits the number of concurrent HTTP requests to the same server to 4.
Without the Semaphore it grows over that number, and causes packet loss.
Actually it works great, and speeds up downloads quite a lot.

Just need to check if the individual created IFirebaseStorage gets freed, when the Task exits..Should, based on your implementation, but need to check on that.

Also unclear, if this effects firebase billing..Creating the Auth several times..Also guess not..

Do you see any issue with such approach ?


Best regards
Thomas

**Reply by SchneiderInfosystems** [reply]
For me this solution is well suited for your described application.

According to Firebase's pricing, this approach should not involve higher costs: https://firebase.google.com/pricing.
Of course, I also only know the effective billing from the publication and our own experiences, and I also don't know what Google plans to do in the future.

**Comment by SchneiderInfosystems** [comment]
I made storage.GetSynchronous thread-safe again. For details check issue https://github.com/SchneiderInfosystems/FB4D/issues/121.
